syntax = "proto3";

package indoors.pipeline.protocol;

enum Channel {
    /* generic platform clock that provides `Clock` */
    CLOCK = 0;

    /* calibrated generic platform accelerometer that provides `Vector3` */
    ACCELEROMETER_CALIBRATED = 10;
    /* calibrated generic platform gyroscope that provides `Vector3` */
    GYROSCOPE_CALIBRATED = 11;
    /* calibrated generic platform magnetometer that provides `Vector3` */
    MAGNETOMETER_CALIBRATED = 12;

    /* uncalibrated generic platform accelerometer that provides `Vector3` */
    ACCELEROMETER_UNCALIBRATED = 20;
    /* uncalibrated generic platform gyroscope that provides `Vector3` */
    GYROSCOPE_UNCALIBRATED = 21;
    /* uncalibrated generic platform magnetometer that provides `Vector3` */
    MAGNETOMETER_UNCALIBRATED = 22;

    /* generic platform accelerometer bias that provides `Vector3` */
    ACCELEROMETER_BIAS = 30;
    /* generic platform gyroscope bias that provides `Vector3` */
    GYROSCOPE_BIAS = 31;
    /* generic platform magnetometer bias that provides `Vector3` */
    MAGNETOMETER_BIAS = 32;

    /* generic platform orientation that provides `Quaternion` */
    ORIENTATION = 40;

    /* generic platform tick that provides `Void` */
    TICK = 100;
}

message Hello {
    string from = 1;
    repeated ChannelHello channels = 2;
}

message Bye {
}

message ChannelHello {
    uint32 channel = 1;
    string annotation = 2;
    Event eventExample = 3;
}

message Event {
    double t = 1;
    uint32 channel = 2;
    oneof _ {
        Hello hello = 3;
        Bye bye = 4;
        Void void = 5;
        bool switch = 6;
        double scalar = 7;
        Clock clock = 8;
        Vector2 vector2 = 9;
        Vector3 vector3 = 10;
        Quaternion quaternion = 11;
        Heading heading = 12;
    }
}

message Void {
}

message Clock {
    double unix_utc = 1;
}

message Vector2 {
    double x = 1;
    double y = 2;
}

message Vector3 {
    double x = 1;
    double y = 2;
    double z = 3;
}

message Quaternion {
    double w = 1;
    double x = 2;
    double y = 3;
    double z = 4;
}

message Heading {
    double north = 1;
    double var_north = 2;
}
